### ISSUE TITLE
Unit test failing in `RecoPixelVertexing/PixelTrackFitting/test/testEigenGPUNoFit.cu`

### ISSUE THREAD

**Issue by ericcano** [cms-sw/cmssw#33797]
The test `TEST INVERSE 5x5` in `RecoPixelVertexing/PixelTrackFitting/test/testEigenGPUNoFit.cu` fails. The GPU inverse is not computed:
```
TEST INVERSE 5x5
Here is the matrix m:
 2.78566  2.00704 0.372114  1.20088   2.2164
 2.00704  2.82538  3.31004  1.35291  2.65503
0.372114  3.31004  0.96297  2.56907  1.58656
 1.20088  1.35291  2.56907  1.56912  2.25059
  2.2164  2.65503  1.58656  2.25059 0.973175
Its inverse is:
  0.194348 -0.0425319  -0.323402 -0.0559519   0.330044
-0.0425319   0.502258   0.235007  -0.686533 -0.0688409
 -0.323402   0.235007  -0.181879  0.0947154   0.172872
-0.0559519  -0.686533  0.0947154   0.660671   0.318141
  0.330044 -0.0688409   0.172872   0.318141   -0.55387
Its GPU inverse is:
0 0 0 0 0
0 0 0 0 0
0 0 0 0 0
0 0 0 0 0
0 0 0 0 0
```
Looking at the compiled code, the corresponding kernel (`kernelInverse5x5`) is an empty function:
```
(cuda-gdb) disassemble
Dump of assembler code for function _Z16kernelInverse5x5PN5Eigen6MatrixIdLi5ELi5ELi0ELi5ELi5EEES2_:
=> 0x0000000000ed4780 <+0>:     MOV R1, c[0x0][0x28]
   0x0000000000ed4790 <+16>:    EXIT
   0x0000000000ed47a0 <+32>:    BRA 0x20
   0x0000000000ed47b0 <+48>:    NOP
   0x0000000000ed47c0 <+64>:    NOP
   0x0000000000ed47d0 <+80>:    NOP
   0x0000000000ed47e0 <+96>:    NOP
   0x0000000000ed47f0 <+112>:   NOP
```
as opposed to `kernelInverse4x4` in spite of having very similar contents.

Compilation gives a warning that might be related:
```
/cvmfs/cms-ib.cern.ch/nweek-02681/slc7_amd64_gcc900/external/eigen/011e0db31d1bed8b7f73662be6d57d9f30fa457a/include/eigen3/Eigen/src/Core/PlainObjectBase.h(602): warning #20011-D: calling a __host__ function("Eigen::PartialPivLU< ::Eigen::Matrix<double, (int)5, (int)5, (int)0, (int)5, (int)5> > ::PartialPivLU< ::Eigen::Matrix<double, (int)5, (int)5, (int)0, (int)5, (int)5> > ") from a __host__ __device__ function("Eigen::PlainObjectBase< ::Eigen::Matrix<double, (int)5, (int)5, (int)0, (int)5, (int)5> > ::PlainObjectBase< ::Eigen::Inverse< ::Eigen::Matrix<double, (int)5, (int)5, (int)0, (int)5, (int)5> > > ") is not allowed
```

**Comment by makortel** [comment]
assign reconstruction, heterogeneous

FYI @VinInn

**Comment by VinInn** [comment]
I would advice to fix all compilation warnings in Eigen.

**Comment by VinInn** [comment]
> That implies hacking Eigen itself. What's the best way to do this? Point scram to a different version?

This is what we (well mostly @fwyzard ) used to do.

**Comment by ericcano** [comment]
Eigen was switched back to upstream on `CMSSW_11_3_0_pre5`. The test was not yet in `CMSSW_11_3_0_pre5`, but might have been developped against the previous Eigen version.
I could test this easily (it works) with:
```
scram setup /cvmfs/cms.cern.ch/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_3_0_pre4/config/toolbox/slc7_amd64_gcc900/tools/selected/eigen.xml
```

**Comment by ericcano** [comment]
The modifications from the branch we previously used (https://github.com/cms-externals/eigen-git-mirror/tree/cms/master/d812f411c3f9) were cherry picked on top of Eigen's HEAD revision (https://gitlab.com/libeigen/eigen/-/commit/391094c50743f28f9174f455661f650bf07e0177). The test now gives a partially correct matrix:
```
TEST INVERSE 4x4
Here is the matrix m:
0.776558  1.40997  2.16143  1.91816
 1.40997  2.69351  2.16311  2.46696
 2.16143  2.16311 0.197677  2.52519
 1.91816  2.46696  2.52519  1.20556
Its inverse is:
  0.53622  -1.30801  0.576264  0.616367
 -1.30801   1.47265 -0.392019 -0.111218
 0.576264 -0.392019 -0.175199   0.25228
 0.616367 -0.111218   0.25228 -0.452057
Its GPU inverse is:
  0.53622  -1.30801  0.576264  0.616367
 -1.30801   1.47265 -0.392019 -0.111218
 0.576264 -0.392019 -0.175199   0.25228
 0.616367 -0.111218   0.25228 -0.452057
TEST INVERSE 5x5
Here is the matrix m:
 2.73042  3.10303  2.27264  1.55492 0.980891
 3.10303  1.57855  2.84141 0.903772   2.3176
 2.27264  2.84141 0.322676  2.58113  1.85137
 1.55492 0.903772  2.58113  1.50039  2.77534
0.980891   2.3176  1.85137  2.77534  1.81274
Its inverse is:
 -1.74061   2.78094 -0.133561  -2.69686   1.65178
  2.78094  -3.77355  0.524643   3.63592  -2.78278
-0.133561  0.524643 -0.512885 -0.460065  0.629694
 -2.69686   3.63592 -0.460065  -3.87164   3.20818
  1.65178  -2.78278  0.629694   3.20818  -2.33925
Its GPU inverse is:
        0  0.322266         0         0         0
 0.908936 -0.619752 -0.189006   3.63592 -0.131644
  0.10331  0.125583 -0.422584 -0.460065  0.294237
-0.703499  0.277659  0.299851  -3.87164  0.385174
        0         0         0   3.20818         0
```

**Comment by ericcano** [comment]
The integration of the previous modifications (plus a small addition) led to no compilation warning related to __host__ functions call. The unit test now runs succesfuly. This version of Eigen is in this branch: https://gitlab.com/ericcano/eigen/-/commits/cms-reimport-issues

We still have this compile-time warning:
```
/data/user/cano/eigen/Eigen/src/Core/functors/AssignmentFunctors.h: In function 'riemannFit::CircleFit riemannFit::circleFit(const M2xN&, riemannFit::Matrix2Nd<N>&, const V4&, riemannFit::VectorNd<N>&, double, bool) [with M2xN = Eigen::Block<Eigen::Matrix<double, 3, 4>, -1, -1, false>; V4 = Eigen::Matrix<double, 4, 1>; int N = 4]':
/data/user/cano/eigen/Eigen/src/Core/functors/AssignmentFunctors.h:92:62: warning: '*((void*)& p3D +16)' is used uninitialized in this function [-Wuninitialized]
   92 |   EIGEN_DEVICE_FUNC EIGEN_STRONG_INLINE void assignCoeff(DstScalar& a, const SrcScalar& b) const { a *= b; }
      |                                                            ~~^~~~
```
and also link time warnings:
```
nvlink warning : Stack size for entry function '_Z16kernelInverse5x5PN5Eigen6MatrixIdLi5ELi5ELi0ELi5ELi5EEES2_' cannot be statically determined (target: sm_60)
nvlink warning : Stack size for entry function '_Z16kernelInverse5x5PN5Eigen6MatrixIdLi5ELi5ELi0ELi5ELi5EEES2_' cannot be statically determined (target: sm_70)
nvlink warning : Stack size for entry function '_Z16kernelInverse5x5PN5Eigen6MatrixIdLi5ELi5ELi0ELi5ELi5EEES2_' cannot be statically determined (target: sm_75)
```

**Comment by fwyzard** [comment]
> link time warnings:
> 
> ```
> nvlink warning : Stack size for entry function '_Z16kernelInverse5x5PN5Eigen6MatrixIdLi5ELi5ELi0ELi5ELi5EEES2_' cannot be statically determined (target: sm_60)
> nvlink warning : Stack size for entry function '_Z16kernelInverse5x5PN5Eigen6MatrixIdLi5ELi5ELi0ELi5ELi5EEES2_' cannot be statically determined (target: sm_70)
> nvlink warning : Stack size for entry function '_Z16kernelInverse5x5PN5Eigen6MatrixIdLi5ELi5ELi0ELi5ELi5EEES2_' cannot be statically determined (target: sm_75)
> ```

That kernel seems to be simply
```c++
__global__ void kernelInverse5x5(Matrix5d *in, Matrix5d *out) { (*out) = in->inverse(); }
```

Is it possible that `inverse()` is recursive for 5x5 matrices ?

**Comment by ericcano** [comment]
This is plausible as from 5x5 (included) a generic method is used while Cramer's rule is used for smaller matrices (https://discord.com/channels/777904510169382942/777904791136370758/845316483633774622).

**Comment by slava77** [comment]
is this still an active issue?

**Comment by iarspider** [comment]
I'm not sure this is 100% the same error, but it fails in the same test: https://cmssdt.cern.ch/SDT/cgi-bin/logreader/slc7_amd64_gcc900/CMSSW_12_0_GPU_X_2021-07-21-2300/unitTestLogs/RecoPixelVertexing/PixelTrackFitting#/3289-3289

**Comment by fwyzard** [comment]
Yes, it's still failing in `CMSSW_12_1_0_pre1`:
```
$CMSSW_BASE/test/cc8_amd64_gcc9/testEigenGPUNoFit_t 
TEST EIGENVALUES
TEST INVERSE 3x3
TEST INVERSE 4x4
TEST INVERSE 5x5
testEigenGPUNoFit_t: /tmp/fwyzard/CMSSW_12_1_0_pre1/src/RecoPixelVertexing/PixelTrackFitting/test/test_common.h:24: bool isEqualFuzzy(C1, C2, double) [with C1 = Eigen::Matrix<double, 5, 5>; C2 = Eigen::Matrix<double, 5, 5>]: Assertion `std::abs(a(i, j) - b(i, j)) < std::min(std::abs(a(i, j)), std::abs(b(i, j))) * epsilon' failed.
Aborted (core dumped)
```

**Comment by ericcano** [comment]
This is fixed in https://github.com/cms-externals/eigen-git-mirror/pull/5. One commit was missing to the latest active version of Eigen.

**Comment by perrotta** [comment]
@ericcano given your comment in https://github.com/cms-sw/cmssw/issues/33797#issuecomment-891921269 I think we can close this issue. Please reopen if I misunderstood
